File: src/python/sql_test.py
import helpers


# pylint: disable=wrong-import-order


########################
# business logic start #
########################


import dataclasses
import json
import typing


@dataclasses.dataclass(frozen=True)
class SQLState:
    state: dict

    def read_table_meta(self, table_name: str) -> dict:
        return self.state.get(table_name, {}).get("metadata", {})

    def read_table_rows(self, table_name: str) -> list[dict]:
        return self.state.get(table_name, {}).get("rows", [])

    def read_information_schema(self) -> list[dict]:
        return [data["metadata"] for data in self.state.values()]

    def write_table_meta(self, table_name: str, data: dict):
        state = self.state
        table = state.get(table_name, {})
        metadata = table.get("metadata", {})
        metadata.update(data)
        table["metadata"] = metadata
        state[table_name] = table
        return self.__class__(state)

    def write_table_rows(self, table_name: str, data: dict):
        state = self.state
        table = state.get(table_name, {})
        rows = table.get("rows", [])
        rows.append(data)
        table["rows"] = rows
        state[table_name] = table
        return self.__class__(state)


class SQLType:
    @staticmethod
    def varchar(data) -> str:
        data_str = str(data).strip()
        if data_str.startswith("'") or data_str.startswith('"'):
            data_str = data_str[1:]
        if data_str.endswith("'") or data_str.endswith('"'):
            data_str = data_str[:-1]
        return data_str

    @staticmethod
    def int(data) -> int:
        return int(data.strip())


sql_type_map = {
    "VARCHAR": SQLType.varchar,
    "INT": SQLType.int,
}


class SQLFunctions:
    @staticmethod
    def create_table(state: SQLState, *args, table_schema="public") -> typing.Tuple[list, SQLState]:
        output: list[dict] = []
        table_name = args[2]

        # get columns
        columns = {}
        columns_str = " ".join(args[3:]).replace("(", "").replace(")", "").strip()
        if columns_str:
            # fmt: off
            columns = {
                column.strip().split(" ")[0]: column.strip().split(" ")[1]
                for column in columns_str.split(",")
            }
            # fmt: on

        if not state.read_table_meta(table_name):
            state = state.write_table_meta(
                table_name,
                {
                    "table_name": table_name,
                    "table_schema": table_schema,
                    "colums": columns,
                },
            )
        return (output, state)

    @staticmethod
    def insert_into(state: SQLState, *args) -> typing.Tuple[list, SQLState]:
        output: list[dict] = []
        table_name = args[2]

        values_index = None
        for i, arg in enumerate(args):
            if arg == "VALUES":
                values_index = i
        if values_index is None:
            raise ValueError("VALUES not found")

        keys = " ".join(args[3:values_index]).replace("(", "").replace(")", "").split(",")
        keys = [key.strip() for key in keys]
        values = " ".join(args[values_index + 1 :]).replace("(", "").replace(")", "").split(",")
        values = [value.strip() for value in values]
        key_value_map = dict(zip(keys, values))

        data = {}
        if metadata := state.read_table_meta(table_name):
            for key, value in key_value_map.items():
                data[key] = sql_type_map[metadata["colums"][key]](value)
            state = state.write_table_rows(table_name, data)

        return (output, state)

    @staticmethod
    def select(state: SQLState, *args) -> typing.Tuple[list, SQLState]:
        output: list[dict] = []

        from_index = None
        where_index = None
        for i, arg in enumerate(args):
            if arg == "FROM":
                from_index = i
            if arg == "WHERE":
                where_index = i
        if from_index is None:
            raise ValueError("FROM not found")

        # get select keys by getting the slice of args before FROM
        select_keys = " ".join(args[1:from_index]).split(",")
        select_keys = [key.strip() for key in select_keys]

        # get where keys by getting the slice of args after WHERE
        from_value = args[from_index + 1]

        # `information_schema.tables` is a special case
        if from_value == "information_schema.tables":
            data = state.read_information_schema()
        else:
            data = state.read_table_rows(from_value)

        output = []
        for datum in data:
            # fmt: off
            output.append({
                key: datum.get(key)
                for key in select_keys
            })
            # fmt: on

        return (output, state)


sql_function_map: dict[str, typing.Callable] = {
    "CREATE TABLE": SQLFunctions.create_table,
    "SELECT": SQLFunctions.select,
    "INSERT INTO": SQLFunctions.insert_into,
}


def run_sql(input_sql: list[str]) -> list[str]:
    output = []
    state = SQLState(state={})

    # remove comments
    input_sql = [line.strip() for line in input_sql if not line.startswith("--")]

    # re-split on semi-colons
    input_sql = " ".join(input_sql).split(";")

    # iterate over each line of sql
    for line in input_sql:
        words = line.split(" ")
        for i in reversed(range(len(words) + 1)):
            key = " ".join(words[:i]).strip()
            if func := sql_function_map.get(key):
                output, state = func(state, *[word for word in words if word])
                break

    return [json.dumps(output)]


######################
# business logic end #
######################

if __name__ == "__main__":
    helpers.run(run_sql)

File: src/ruby/sql_test.rb
# frozen_string_literal: true

require 'json'

########################
# business logic start #
########################

class SQLState
  attr_reader :state

  def initialize(state = {})
    @state = state
  end

  def read_table_meta(table_name)
    @state[table_name] ? @state[table_name]['metadata'] || {} : {}
  end

  def read_table_rows(table_name)
    @state[table_name] ? @state[table_name]['rows'] || [] : []
  end

  def read_information_schema
    @state.values.map { |data| data['metadata'] }
  end

  def write_table_meta(table_name, data)
    @state[table_name] ||= {}
    @state[table_name]['metadata'] ||= {}
    @state[table_name]['metadata'].merge!(data)
    self.class.new(@state)
  end

  def write_table_rows(table_name, data)
    @state[table_name] ||= {}
    @state[table_name]['rows'] ||= []
    @state[table_name]['rows'] << data
    self.class.new(@state)
  end
end

module SQLType
  def self.varchar(data)
    data.to_s.strip.sub(/^['"]/, '').sub(/['"]$/, '')
  end

  def self.int(data)
    data.to_i
  end
end

module SQLFunctions
  def self.create_table(state, *args, table_schema: 'public')
    output = []
    table_name = args[2]

    columns = {}
    columns_str = args[3..].join(' ').gsub(/[()]/, '').strip
    unless columns_str.empty?
      columns = Hash[columns_str.split(',').map { |column| column.strip.split(' ') }]
    end

    if state.read_table_meta(table_name).empty?
      metadata = {
        'table_name' => table_name,
        'table_schema' => table_schema,
        'columns' => columns
      }
      state = state.write_table_meta(table_name, metadata)
    end

    [output, state]
  end

  def self.insert_into(state, *args)
    output = []
    table_name = args[2]

    sql_type_map = {
      'VARCHAR' => SQLType.method(:varchar),
      'INT' => SQLType.method(:int)
    }

    values_index = args.index('VALUES')
    raise 'VALUES not found' if values_index.nil?

    keys = args[3...values_index].join(' ').gsub(/[()]/, '').split(',').map(&:strip)
    values = args[values_index + 1..].join(' ').gsub(/[()]/, '').split(',').map(&:strip)
    key_value_map = Hash[keys.zip(values)]

    metadata = state.read_table_meta(table_name)
    row = {}
    key_value_map.each do |key, value|
      row[key] = sql_type_map[metadata['columns'][key]].call(value)
    end
    state = state.write_table_rows(table_name, row)

    [output, state]
  end

  def self.select(state, *args)
    output = []

    from_index = args.index('FROM')
    # where_index = args.index('WHERE')

    raise 'FROM not found' if from_index.nil?

    select_keys = args[1...from_index].join(' ').split(',').map {|s| s.gsub(/[()]/, '')}.map(&:strip)
    from_value = args[from_index + 1]

    data = if from_value == 'information_schema.tables'
             state.read_information_schema
           else
             state.read_table_rows(from_value)
           end

    data.each do |datum|
      output << select_keys.map { |key| [key, datum[key]] }.to_h
    end

    [output, state]
  end
end

def run_sql(input_sql)
  output = []
  state = SQLState.new

  input_sql = input_sql.map(&:strip).reject { |line| line.start_with?('--') }
  input_sql = input_sql.join(' ').split(';')

  sql_function_map = {
    'CREATE TABLE' => SQLFunctions.method(:create_table),
    'SELECT' => SQLFunctions.method(:select),
    'INSERT INTO' => SQLFunctions.method(:insert_into)
  }

  input_sql.each do |line|
    words = line.split(' ')
    words.each_index do |i|
      key = words[0..i].join(' ').strip
      if sql_function_map.key?(key)
        output, state = sql_function_map[key].call(state, *words.reject(&:empty?))
        break
      end
    end
  end

  [output.to_json]
end

######################
# business logic end #
######################

if __FILE__ == $PROGRAM_NAME
  input_file_path = ENV['INPUT_PATH']
  input_file = File.readlines(input_file_path)

  sorted_data = run_sql(input_file)

  output_file_path = ENV['OUTPUT_PATH']
  File.open(output_file_path, 'wb') { |f| f.write(sorted_data.join('')) }
end
